packages/block-editor/src/components/block-list/insertion-point.js

```diff
@@ -1,74 +1,119 @@
-/**
- * External dependencies
- */
-import classnames from 'classnames';
-
 /**
  * WordPress dependencies
  */
-import { useState } from '@wordpress/element';
 import { useSelect } from '@wordpress/data';
+import { useState } from '@wordpress/element';
+import { Popover } from '@wordpress/components';
 
 /**
  * Internal dependencies
  */
 import Inserter from '../inserter';
 
-export default function BlockInsertionPoint( { rootClientId, clientId } ) {
-	const [ isInserterFocused, setInserterFocused ] = useState( false );
+function Indicator( { clientId } ) {
 	const showInsertionPoint = useSelect( ( select ) => {
 		const {
 			getBlockIndex,
 			getBlockInsertionPoint,
 			isBlockInsertionPointVisible,
+			getBlockRootClientId,
 		} = select( 'core/block-editor' );
+		const rootClientId = getBlockRootClientId( clientId );
 		const blockIndex = getBlockIndex( clientId, rootClientId );
 		const insertionPoint = getBlockInsertionPoint();
 		return (
 			isBlockInsertionPointVisible() &&
 			insertionPoint.index === blockIndex &&
 			insertionPoint.rootClientId === rootClientId
 		);
-	}, [ clientId, rootClientId ] );
+	}, [ clientId ] );
 
-	function onFocus( event ) {
-		// Stop propagation of the focus event to avoid selecting the current
-		// block while inserting a new block, as it is not relevant to sibling
-		// insertion and conflicts with contextual toolbar placement.
-		event.stopPropagation();
-		setInserterFocused( true );
+	if ( ! showInsertionPoint ) {
+		return null;
 	}
 
-	function onBlur() {
-		setInserterFocused( false );
+	return <div className="block-editor-block-list__insertion-point-indicator" />;
+}
+
+export default function InsertionPoint( {
+	className,
+	isMultiSelecting,
+	selectedBlockClientId,
+	children,
+} ) {
+	const [ isInserterShown, setIsInserterShown ] = useState( false );
+	const [ isInserterForced, setIsInserterForced ] = useState( false );
+	const [ inserterElement, setInserterElement ] = useState( null );
+	const [ inserterClientId, setInserterClientId ] = useState( null );
+
+	function onMouseMove( event ) {
+		if ( event.target.className !== className ) {
+			if ( isInserterShown ) {
+				setIsInserterShown( false );
+			}
+			return;
+		}
+
+		const rect = event.target.getBoundingClientRect();
+		const offset = event.clientY - rect.top;
+		const element = Array.from( event.target.children ).find( ( blockEl ) => {
+			return blockEl.offsetTop > offset;
+		} );
+
+		if ( ! element ) {
+			return;
+		}
+
+		const clientId = element.id.slice( 'block-'.length );
+
+		if ( ! clientId || clientId === selectedBlockClientId ) {
+			return;
+		}
+
+		const elementRect = element.getBoundingClientRect();
+
+		if ( event.clientX > elementRect.right || event.clientX < elementRect.left ) {
+			if ( isInserterShown ) {
+				setIsInserterShown( false );
+			}
+			return;
+		}
+
+		setIsInserterShown( true );
+		setInserterElement( element );
+		setInserterClientId( clientId );
```

Comment:
While it might not be too much an issue since `setState` won't cause a render unless the state actually changes, I note (by placing a [logpoint](https://developers.google.com/web/updates/2019/01/devtools#logpoints)) that this gets called upwards of ~15 times every time a cursor moves within a sibling inserter. Is it something where (even if we don't strictly need to), we could abort the logic when determining that nothing needs to change?

Reply:
Would it make much difference, you think? I thought the only thing we'd avoid is an extra function call.

Reply:
> Would it make much difference, you think? I thought the only thing we'd avoid is an extra function call.

It's hard to say. I'm not too sure what all React does internally to decide how to handle `setState`. Depending if we could abort any earlier, it could be an opportunity to [avoid reflow from one of the `getBoundingClientRect`](https://gist.github.com/paulirish/5d52fb081b3570c81e3a)? I'm not sure of the implementation here to know whether that's even possible; I'd guess by the fact there are already some `return;` that it's probably not?